### PR TITLE
std.math.poly: optimisation for LDC.

### PR DESCRIPTION
**Pull request by 9il**
Many functions from `std.math` and `std.mathspecial` use `poly`, so
this PR affects significant part of math.

See also: http://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -6339,15 +6339,31 @@ unittest {
...
+        {
+            alias F = floatTraits!(typeof(return));
+            static if(F.realFormat == RealFormat.ieeeSingle || RealFormat.ieeeDouble)
+            {
+                r = llvm_fmuladd(r, x, A[i]);
+            }
+            else
+            {
+                r *= x;
+                r += A[i];
```
Comment (John-Colvin) [inline std/math.d]: why not ```r = r * x + A[i]```?

**Comment by John-Colvin** [comment]
LGTM, as long as it doesn't break any of the math/mathspecial tests when using LDC?

**Comment by ibuclaw** [comment]
I don't think that we should be allowing compiler-specific things into phobos (I/we make the same complaint with dmd-specific code too).

File: std/math.d (hunk 1 above)
Comment (ibuclaw) [inline std/math.d]: Does this even optimized anything?  I thought that ``fmadd`` was only for vectors.

Comment (9il) [inline std/math.d]: Yes, all vectors opcodes has version for single floating point numbers.

Comment (9il) [inline std/math.d]: @John-Colvin it looks more understandable to me

Comment (ibuclaw) [inline std/math.d]: Ah ha, I wasn't looking hard enough.  I might suggest using ``core.stdc.math.fma`` and ``core.stdc.math.fmaf`` instead for all  (also, GDC recognizes all ``core.stdc.*`` declarations as builtins).

**Comment by ibuclaw** [comment]
Any hard numbers proving this generates better code?

**Comment by 9il** [comment]
> I don't think that we should be allowing compiler-specific things into phobos (I/we make the same complaint with dmd-specific code too).

Looks reasonable, then I should move this code to LDC fork.

> Any hard numbers proving this generates better code?

Yes: *The ‘llvm.fmuladd‘ intrinsic functions represent multiply-add expressions that can be fused if the code generator determines that (a) the target instruction set has support for a fused operation, and (b) that the fused operation is more efficient than the equivalent, separate pair of mul and add instructions.*

> I might suggest using core.stdc.math.fma and core.stdc.math.fmaf instead (also, GDC recognizes all core.stdc.* declarations as builtins).

Is `fma` *always* fused or only if (a) and (b) like `llvm.fmuladd `?

**Comment by 9il** [comment]
> Please don't. If core.stdc.math.fma cannot be used (without looking at the docs, I suspect that they guarantee rounding only once), we should add an appropriate intrinsic to core.math. Phobos should be shared across all the compilers.

Agreed.

**Comment by 9il** [comment]
Other reason to add intrinsic to `core.math` is that LLVM inline assembler is not CTFEable.

**Comment by ibuclaw** [comment]
> Other reason to add intrinsic to core.math is that LLVM inline assembler is not CTFEable.

Not that adding it as an intrinsic will magically make the LLVM inline assembler CTFEable either...

**Pull request closed without merging**